### PR TITLE
SigMFFile class is now an iterable for accessing data via memmap

### PR DESCRIPTION
**Pull request by gmabey**
It also now has .ndim and .shape members and has a __getitem__() method.

The purpose in this patch is to better facilitate the accessing of very large data sets through the use of `numpy.memmap()`.  The complications arise when handling types supported by SigMF but not by numpy, like fixed-precision complex samples.  The proposed solution still provides access to the data on disk through the use of `._memmap` (expert mode) but when using the bracket operator, useful datatypes are always returned.

I would like to think that this new approach could someday make the `readsamples()` method (in its current form) obsolete, but that's premature even if it were the right thing to do

Test with this code:
```python
#!/usr/bin/env python3

import json
import numpy
import sigmf
import warnings

warnings.filterwarnings("always")

meta = {
    "global": {
        "core:author": "Glen M",
        "core:datatype": "ri16_le",
        "core:license": "https://creativecommons.org/licenses/by-sa/4.0/",
        "core:num_channels": 2,
        "core:sample_rate": 48000,
        "core:version": "1.0.0"
    },
    "captures": [
        {
            "core:datetime": "2021-06-18T23:17:51.163959Z",
            "core:sample_start": 0
        }
    ] }

NUM_ROWS = 5

for dt in "ri16_le", "ci16_le", "rf32_le", "rf64_le", "cf32_le", "cf64_le":
    meta["global"]["core:datatype"] = dt
    for num_chan in 1,3:
        meta["global"]["core:num_channels"] = num_chan
        base_filename = dt + '_' + str(num_chan)
        meta_filename = base_filename + '.sigmf-meta'
        meta_fid = open(meta_filename, 'w')
        json.dump(meta, meta_fid)
        meta_fid.close()
        a = numpy.arange(NUM_ROWS * num_chan * (2 if 'c' in dt else 1))
        if 'i16' in dt:
            b = a.astype(numpy.int16)
        elif 'f32' in dt:
            b = a.astype(numpy.float32)
        elif 'f64' in dt:
            b = a.astype(numpy.float64)
        else:
            raise ValueError('whoops')

        b.tofile(base_filename + '.sigmf-data')

        f = sigmf.sigmffile.fromfile(meta_filename)
        print(f'{dt} with {num_chan} channels: .ndim = {f.ndim} .shape = {f.shape}')
        print(f._memmap[:])
        print(f[:])
        print([i for i in f])
        print()
```

### CODE REVIEW

**Comment by bhilburn** [comment]
@gmabey - This looks excellent! I'm marking this as non-required for the v1.0.0 tag, but I'm eager to get it into the v1.x branch as soon as it's reviewed and ready 🙂

@jacobagilbert @Teque5 - would love your reviews here

**Comment by gmabey** [comment]
The first commit already had support in it for `complex128`, so with minimal touch-ups, this Fixes #196 so it may be of interest to @jacobagilbert @777arc @jsallay @sandersa-nist

**Comment by gmabey** [comment]
I've now added another class that allows for accessing data without un-tar-ing an archive.  The intent is to address the case where an archive is so large that it is onerous/infeasible/obnoxious to unpack it.  Please consider exercising this patch with this script:
```python
#!/usr/bin/env python3

import json
import numpy
import sigmf
import warnings

warnings.filterwarnings("always")

global_info = {
        "core:author": "Glen M",
        "core:datatype": "ri16_le",
        "core:license": "https://creativecommons.org/licenses/by-sa/4.0/",
        "core:num_channels": 2,
        "core:sample_rate": 48000,
        "core:version": "1.0.0"
    }
capture_info = {
        "core:datetime": "2021-06-18T23:17:51.163959Z",
        "core:sample_start": 0
    }

NUM_ROWS = 5

for dt in "ri16_le", "ci16_le", "rf32_le", "rf64_le", "cf32_le", "cf64_le":
    global_info["core:datatype"] = dt
    for num_chan in 1,3:
        global_info["core:num_channels"] = num_chan
        base_filename = dt + '_' + str(num_chan)
        archive_filename = base_filename + '.sigmf'
        data_filename = base_filename + '.sigmf-data'

        a = numpy.arange(NUM_ROWS * num_chan * (2 if 'c' in dt else 1))
        if 'i16' in dt:
            b = a.astype(numpy.int16)
        elif 'f32' in dt:
            b = a.astype(numpy.float32)
        elif 'f64' in dt:
            b = a.astype(numpy.float64)
        else:
            raise ValueError('whoops')

        b.tofile(data_filename)
        meta = sigmf.SigMFFile(data_file=data_filename, global_info=global_info)
        meta.add_capture(0, metadata=capture_info)
        meta.tofile(archive_filename, toarchive=True)

        archi = sigmf.SigMFArchiveReader(archive_filename, skip_checksum=True)
        print(f'{dt} with {num_chan} channels: .ndim = {archi.ndim} .shape = {archi.shape}')
        print(archi.sigmffile._memmap[:])
        print(archi[:])
        print([i for i in archi])
        print()
```

File: sigmf/sigmf_hash.py
Comment (gmabey) [inline sigmf/sigmf_hash.py:32]: I'm sad to see this concise line become so much more bloated ...

File: sigmf/sigmffile.py
Comment (gmabey) [inline sigmf/sigmffile.py:219]: If `new_global` is a dict that is used to create multiple `.sigmf` files, then we probably don't want things like the `sha512` to be reflected in the dict at the level that invoked (for example) `SigMFFile.__init__()`

**Review by jacobagilbert — CHANGES_REQUESTED** [review]
There are three mostly separate things going on here that are complicating review. I would recommend we get these split up so that they can be reviewed effectively:

1. Using numpy.memmap - this is a great idea, and I'd like to have a discussion about entirely switching to this for all dataset accesses.
2. Adding double precision float support - I think we will end up adding support but this should be in a separate PR.
3. Allowing access to compressed datafiles - also a good idea and possibly the only way we should be doing this. Please open another PR for this.

Let's get this refocused around #1 and figure out how to make that happen.

**Comment by gmabey** [comment]
> 3\. compressed datafiles

Do you mean tarball'ed archives?  compression is different, right?

**Comment by gmabey** [comment]
Ok, I'm somewhat of a klutz WRT git, but I think I've separated those two other commits into their own branches now.

**Review by Teque5 — COMMENTED** [review]
I think this looks good, but there should be

1. At least a basic test added for this.
2. Increment version micro

Let me see what I can do...

**Comment by Teque5** [comment]
@gmabey I submitted a PR to your `iterable` branch with my suggested improvements. After you merge that I vote this is ready. I incremented the *micro* too.

**Comment by gmabey** [comment]
Slightly off topic from this PR, but shouldn't the `SigMFFile` class be named `SigMFRecording` instead?

**Comment by Teque5** [comment]
> Slightly off topic from this PR, but shouldn't the `SigMFFile` class be named `SigMFRecording` instead?

You might be right about this, but it might be even more correct to call the top level object `SigMF`, so it's intuitive to just use things like `sigmf.SigMF.GLOBAL_KEY`.

Since that kind of change would break the API, we should decide how the versioning for the software is differentiated from the versioning for the spec before we do that PR.

**Comment by jacobagilbert** [comment]
This would be a 2.x change.

**Review by Teque5 — APPROVED** [review]
@gmabey incorported my changes and I added testing, so this looks great now.

**Comment by gmabey** [comment]
@jacobagilbert any chance I could get you to please review this PR? Also -- is there an open question as to whether using memmap is ok to always use?

Hmm -- looks like GitHub believes that @jacobagilbert requested changes, but I can't tell what was actually requested. Any insight into this would be appreciated.

**Review by jacobagilbert — APPROVED** [review]
This looks good to me @gmabey. I cannot see what changes i requested either for some reason, but its been so long I went ahead and had a complete re-review.

**Pull request merged**
